ui/src/app/shared/services/workflows-service.ts

```diff
@@ -118,27 +120,66 @@ export class WorkflowsService {
             .then(res => res.body as Workflow);
     }
 
+    public getContainerLogsFromCluster(workflow: Workflow, nodeId: string, container: string): Observable<string> {
+        const podLogsURL = `api/v1/workflows/${workflow.metadata.namespace}/${workflow.metadata.name}/${nodeId}/log?logOptions.container=${container}&logOptions.follow=true`;
+        return requests
+            .loadEventSource(podLogsURL)
+            .map(line => JSON.parse(line).result.content)
+            .filter(isString)
+            .catch(() => {
+                // When an error occurs on an observable, RxJS is hard-coded to unsubscribe from the stream.  In the case
+                // that the connection to the server was interrupted while the node is still pending or running, this is not
+                // correct since we actually want the EventSource to re-connect and continue streaming logs.  In the event
+                // that the pod has completed, then we want to allow the unsubscribe to happen since no additional logs exist.
+                return Observable.fromPromise(this.isWorkflowNodePendingOrRunning(workflow, nodeId)).switchMap(isPendingOrRunning => {
+                    if (isPendingOrRunning) {
+                        return this.getContainerLogsFromCluster(workflow, nodeId, container);
+                    }
+
+                    // If our workflow is completed, then simply complete the Observable since nothing else
+                    // should be omitted
+                    return Observable.empty();
+                });
+            });
+    }
+
+    public async isWorkflowNodePendingOrRunning(workflow: Workflow, nodeId: string) {
+        if (this.isNodePendingOrRunning(workflow.status.nodes[nodeId])) {
+            return false;
+        }
+
+        // TODO: What happens if we're offline when we make this request?  We should figure out what errors we should
+        // retry on, and what errors we should allow to occur, then transform this to an Observable that implements
+        // all of that
+        const updatedWorkflow = await this.get(workflow.metadata.namespace, workflow.metadata.name);
```

Comment:
Do we want to retry indefinitely in the event that we receive a network failure?  It's relatively straightforward to implement, but want to confirm first

Reply:
No. Just once. Provide a way for the user to refresh, e.g. use `<ErrorNotice onReload={...}/>`